### PR TITLE
Avoid PyThread in `stream.pyx`

### PR DESCRIPTION
**Pull request by niboshi**
Fixes a part of #1940.

This PR also includes some refactoring of thread-local storage. The purpose is to reduce the number of runtime lookups for thread-local variables.

### CODE REVIEW

File: cupy/cuda/stream.pyx [hunk 1]
```diff
@@ -173,24 +190,27 @@ class Stream(object):
...
+        tls.prev_stream_ref_stack.append(prev_stream_ref)
+        tls.set_current_stream(self)
         return self
 
     def __exit__(self, *args):
-        prev_stream_ref = _thread_local.prev_stream_ref_stack.pop()
-        _set_current_stream(prev_stream_ref())
+        tls = _StreamThreadLocal.get()
+        prev_stream_ref = tls.prev_stream_ref_stack.pop()
+        tls.set_current_stream(prev_stream_ref())
```
Comment (beam2d) [inline cupy/cuda/stream.pyx]: Similarly as above, here a strong ref is created from a weak ref and then a new weak ref is created from it. Can we remove the redundancy?

**Comment by niboshi** [comment]
Fixed as suggested in the review.
Also I removed checks of `stream` arguments for `None`, as they are never `None`.

**Comment by niboshi** [comment]
Reopened the PR, because the difference between `PyThread` and `threading.local` is not so large in my benchmark:
```

**Comment by niboshi** [comment]
@beam2d @okuta 
How do you think?
I think we can't justify the use of deprecated (and cryptic) API for 30ns.

File: cupy/cuda/stream.pyx [hunk 2]
```diff
@@ -1,21 +1,58 @@
...
-from cpython cimport pythread
 import threading
 import weakref
 
 
 cdef object _thread_local = threading.local()
-cdef int _current_stream_key = pythread.PyThread_create_key()
+
+
+cdef class _StreamThreadLocal(object):
```
[the remark this answers is not in the document]
Comment (niboshi) [reply cupy/cuda/stream.pyx]: Fixed.

File: cupy/cuda/stream.pyx [hunk 3]
```diff
@@ -159,9 +178,10 @@ class Stream(object):
 
     def __del__(self):
         if self.ptr:
-            current_ptr = get_current_stream_ptr()
-            if self.ptr == current_ptr:
-                _set_current_stream(self.null)
+            tls = _StreamThreadLocal.get()
+            current_ptr = tls.get_current_stream_ptr()
+            if self.ptr == <size_t>current_ptr:
```
[the remark this answers is not in the document]
Comment (niboshi) [reply cupy/cuda/stream.pyx]: Fixed.

**Comment by niboshi** [comment]
Fixed.

**Comment by okuta** [comment]
jenkins, test this please.

**Comment by okuta** [comment]
LGTM!

**Pull request merged**